### PR TITLE
chore(sql): implement PgWire Authenticator which controls the network socket

### PR DESCRIPTION
**Pull request by jerrinot**
The Authenticator owns the network socket until authentication is finished - either successfully or a failure.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 302 / 325 (92.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|248|264|93.94%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|27|29|93.10%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 300 / 325 (92.31%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|248|264|93.94%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|27|29|93.10%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 297 / 320 (92.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|242|258|93.80%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|27|29|93.10%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 321 / 349 (91.98%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/security/ReadOnlySecurityContextFactory.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/ReadOnlyUsersAwareSecurityContextFactory.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|26|28|92.86%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|240|256|93.75%|
|:large_blue_circle:|io/questdb/cairo/security/AllowAllSecurityContextFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticUserDatabase.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 322 / 350 (92.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/security/ReadOnlySecurityContextFactory.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/ReadOnlyUsersAwareSecurityContextFactory.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|26|28|92.86%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|240|256|93.75%|
|:large_blue_circle:|io/questdb/cairo/security/AllowAllSecurityContextFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticUserDatabase.java|5|5|100.00%|

File: core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+    private long recvBufStart;
+    private long recvBufWritePos;
+    private long sendBufEnd;
+    private long sendBufReadPos;
+    private long sendBufStart;
+    private long sendBufWritePos;
+    private State state = State.EXPECT_INIT_MESSAGE;
+    private CharSequence username;
+
+    public CleartextPasswordPgWireAuthenticator(NetworkFacade nf, PGWireConfiguration configuration,
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java:77]: nit: arg formatting

File: core/src/main/java/io/questdb/cairo/security/SecurityContextFactory.java [hunk 2]
```diff
@@ -27,11 +27,12 @@
 import io.questdb.cairo.SecurityContext;
 
 public interface SecurityContextFactory {
-    SecurityContext getInstance(CharSequence principal, boolean readOnlyContext);
+    int HTTP = 0;
+    int PGWIRE = 1;
+    int ILP = 2;
+    int TEST_ONLY = -1;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/security/SecurityContextFactory.java:33]: unused

File: core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+        this.sink = new ResponseSink();
+        this.serverVersion = configuration.getServerVersion();
+        this.circuitBreaker = circuitBreaker;
+        this.optionsListener = optionsListener;
+    }
+
+    public CleartextPasswordPgWireAuthenticator(NetworkFacade nf, PGWireConfiguration configuration,
+                                                int circuitBreakerId, NetworkSqlExecutionCircuitBreaker circuitBreaker, CircuitBreakerRegistry registry,
+                                                OptionsListener optionsListener) {
+        this(nf, configuration, circuitBreakerId, circuitBreaker, registry, optionsListener, new StaticUserDatabase(configuration));
```
[the remark this answers is not in the document]
Comment (jerrinot) [reply core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java:98]: good one, thanks!

File: core/src/main/java/io/questdb/cairo/security/SecurityContextFactory.java (hunk 2 above)
Comment (jerrinot) [reply core/src/main/java/io/questdb/cairo/security/SecurityContextFactory.java:33]: it's used from the EE project

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 343 / 372 (92.20%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/Authenticator.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/security/ReadOnlySecurityContextFactory.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/ReadOnlyUsersAwareSecurityContextFactory.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|246|262|93.89%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|31|33|93.94%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/security/AllowAllSecurityContextFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticUserDatabase.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 349 / 379 (92.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/Authenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/security/ReadOnlySecurityContextFactory.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/ReadOnlyUsersAwareSecurityContextFactory.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|250|266|93.98%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/security/AllowAllSecurityContextFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticUserDatabase.java|5|5|100.00%|

File: core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java [hunk 4]
```diff
@@ -7000,22 +6965,25 @@ public void testSelectBindVarsAsyncPG() throws Exception {
...
-                "<3100000004740000000a0001000002bd540000001a00016100000000000001000002bd0008ffffffff0000\n" +
-                ">420000002e005f5f6173796e6370675f73746d745f315f5f00000100010001000000083fd999999999999a00010001450000000900000000005300000004\n" +
-                "<320000000444000000120001000000083fe6666666666666430000000d53454c4543542031005a0000000549\n" +
-                ">5800000004\n";
-        assertHexScript(
-                getFragmentedSendFacade(),
-                script,
-                getHexPgWireConfig()
-        );
+        for (int i = 0; i < 1; i++) {
```
[the remark this answers is not in the document]
Comment (jerrinot) [reply core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java:6968]: ah, yes, from one of the _a hex-test-failing-what-the-hell-is-going-on_ session :) 
thanks!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 354 / 384 (92.19%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/Authenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/security/ReadOnlySecurityContextFactory.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|10|11|90.91%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/ReadOnlyUsersAwareSecurityContextFactory.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|248|264|93.94%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/security/AllowAllSecurityContextFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticPgWireAuthenticationFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticUserDatabase.java|5|5|100.00%|

File: core/src/main/java/io/questdb/cutlass/auth/Authenticator.java [hunk 5]
```diff
@@ -24,19 +24,35 @@
...
 package io.questdb.cutlass.auth;
 
-import io.questdb.cutlass.line.tcp.LineTcpConnectionContext;
-import io.questdb.cutlass.line.tcp.NetworkIOJob;
+import java.io.Closeable;
+import java.io.IOException;
 
-public interface Authenticator {
-    boolean isAuthenticated();
+public interface Authenticator extends Closeable {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/auth/Authenticator.java:30]: we have `QueitClosable` to not include IOException

File: core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+        while (lo < msgLimit - 1) {
+            final long nameLo = lo;
+            final long nameHi = PGConnectionContext.getStringLength(lo, msgLimit, "malformed property name");
+            final long valueLo = nameHi + 1;
+            final long valueHi = PGConnectionContext.getStringLength(valueLo, msgLimit, "malformed property value");
+            lo = valueHi + 1;
+
+            // store user
+            dbcs.of(nameLo, nameHi);
+            if (Chars.equals(dbcs, "user")) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java:407]: nit: have a look at `CairoKeywords` there is a faster method to check for ascii constant

File: core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+                CharacterStoreEntry e = characterStore.newEntry();
+                e.put(dbcs.of(valueLo, valueHi));
+                this.username = e.toImmutable();
+            }
+            boolean parsed = true;
+            if (Chars.equals(dbcs, "options")) {
+                dbcs.of(valueLo, valueHi);
+                if (Chars.startsWith(dbcs, "-c statement_timeout=")) {
+                    try {
+                        long statementTimeout = Numbers.parseLong(dbcs.of(valueLo + "-c statement_timeout=".length(), valueHi));
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java:417]: this is some strange inlining, may be tidy up?

File: core/src/main/java/io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java [hunk 8]
```diff
@@ -24,11 +24,19 @@
...
 
     @Override
-    public ClearTextPgWireAuthenticator getPgWireAuthenticator(PGConnectionContext.ResponseAsciiSink responseAsciiSink, PGWireConfiguration configuration) {
-        return new ClearTextPgWireAuthenticator(responseAsciiSink, configuration);
+    public Authenticator getPgWireAuthenticator(NetworkFacade nf,
+                                                PGWireConfiguration configuration,
+                                                NetworkSqlExecutionCircuitBreaker circuitBreaker,
+                                                CircuitBreakerRegistry registry,
+                                                OptionsListener optionsListener) {
+        return new CleartextPasswordPgWireAuthenticator(nf, configuration, circuitBreaker, registry, optionsListener, new StaticUserDatabase(configuration));
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java:40]: there is no need to create static database more than once

File: core/src/main/java/io/questdb/cutlass/pgwire/PGConnectionContext.java [hunk 9]
```diff
@@ -29,8 +29,11 @@
 import io.questdb.cairo.*;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/pgwire/PGConnectionContext.java:29]: there are new warnings in this file

**Comment by bluestreak01** [comment]
very solid effort on balance, minor tidy ups left! Good job on consolidating authenticator interfaces!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 355 / 385 (92.21%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/Authenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/security/ReadOnlySecurityContextFactory.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|10|11|90.91%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/ReadOnlyUsersAwareSecurityContextFactory.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|249|265|93.96%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|33|35|94.29%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/security/AllowAllSecurityContextFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticPgWireAuthenticationFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticUserDatabase.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 395 / 427 (92.51%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/auth/Authenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cutlass/auth/AnonymousAuthenticator.java|0|2|00.00%|
|:large_blue_circle:|io/questdb/cairo/security/ReadOnlySecurityContextFactory.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|10|11|90.91%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/ReadOnlyUsersAwareSecurityContextFactory.java|9|10|90.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|34|37|91.89%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/CleartextPasswordPgWireAuthenticator.java|247|263|93.92%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGKeywords.java|41|42|97.62%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGWireServer.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/security/AllowAllSecurityContextFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/auth/EllipticCurveAuthenticator.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticPgWireAuthenticationFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/StaticUserDatabase.java|5|5|100.00%|

File: core/src/main/java/io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java (hunk 8 above)
Comment (jerrinot) [reply core/src/main/java/io/questdb/cutlass/pgwire/DefaultPgWireAuthenticationFactory.java]: @bluestreak01 looks like this is the only non-resolved issue. I know you were planning to look at this last night. any success? do you want me to look at this? FYI: it's only relevant with the DefaultServerConfiguration, not when property-based config is used.

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**